File: train.js
#!/usr/bin/env node
// One-time offline training script.
// Run: node train.js
// Outputs: model/model.json + model/*.bin
//
// Requires:  npm install @tensorflow/tfjs-node fs path
// Data:      data/emnist-letters-train-images-idx3-ubyte
//            data/emnist-letters-train-labels-idx1-ubyte

const tf   = require('@tensorflow/tfjs-node');
const fs   = require('fs');
const path = require('path');

const EMNIST_IMAGES = path.join(__dirname, 'data', 'emnist-letters-train-images-idx3-ubyte');
const EMNIST_LABELS = path.join(__dirname, 'data', 'emnist-letters-train-labels-idx1-ubyte');
const OUT_DIR       = path.join(__dirname, 'model');
const TRAINING_N    = 30000;
const EPOCHS        = 20;

function parseIDX3(buf) {
  const v = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
  if (v.getUint32(0) !== 0x00000803) throw new Error('Bad IDX3 magic');
  return { data: buf.slice(16), count: v.getUint32(4), rows: v.getUint32(8), cols: v.getUint32(12) };
}

function parseIDX1(buf) {
  const v = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
  if (v.getUint32(0) !== 0x00000801) throw new Error('Bad IDX1 magic');
  return { data: buf.slice(8), count: v.getUint32(4) };
}

function fixOrientation(src, rows, cols) {
  const dst = new Uint8Array(rows * cols);
  for (let r = 0; r < rows; r++)
    for (let c = 0; c < cols; c++)
      dst[r * cols + c] = src[c * cols + r];
  return dst;
}

function buildModel() {
  const m = tf.sequential();
  m.add(tf.layers.conv2d({ inputShape: [28, 28, 1], filters: 32, kernelSize: 3, activation: 'relu', padding: 'same' }));
  m.add(tf.layers.maxPooling2d({ poolSize: 2 }));
  m.add(tf.layers.conv2d({ filters: 64, kernelSize: 3, activation: 'relu', padding: 'same' }));
  m.add(tf.layers.maxPooling2d({ poolSize: 2 }));
  m.add(tf.layers.flatten());
  m.add(tf.layers.dense({ units: 128, activation: 'relu' }));
  m.add(tf.layers.dropout({ rate: 0.3 }));
  m.add(tf.layers.dense({ units: 26, activation: 'softmax' }));
  m.compile({ optimizer: 'adam', loss: 'sparseCategoricalCrossentropy', metrics: ['acc'] });
  return m;
}

async function main() {
  console.log('Loading EMNIST data…');
  const imgBuf = fs.readFileSync(EMNIST_IMAGES);
  const lblBuf = fs.readFileSync(EMNIST_LABELS);

  const imgs = parseIDX3(imgBuf);
  const lbls = parseIDX1(lblBuf);
  const n    = Math.min(imgs.count, TRAINING_N);
  const { rows, cols } = imgs;

  console.log(`Building tensors from ${n} samples…`);
  const flat = new Float32Array(n * rows * cols);
  for (let i = 0; i < n; i++) {
    const raw   = imgs.data.slice(i * rows * cols, (i + 1) * rows * cols);
    const fixed = fixOrientation(raw, rows, cols);
    for (let p = 0; p < fixed.length; p++)
      flat[i * rows * cols + p] = fixed[p] / 255;
  }
  const xs = tf.tensor4d(flat, [n, rows, cols, 1]);
  const ys = tf.tensor1d(Array.from(lbls.data.slice(0, n)).map(v => v - 1), 'float32');

  const m = buildModel();

  let bestValAcc = -Infinity, patience = 0;
  console.log('Training…');
  await m.fit(xs, ys, {
    epochs: EPOCHS,
    batchSize: 512,
    validationSplit: 0.1,
    callbacks: {
      onEpochEnd: (epoch, logs) => {
        console.log(`  Epoch ${epoch + 1}: acc=${(logs.acc * 100).toFixed(1)}%  val=${(logs.val_acc * 100).toFixed(1)}%`);
        if (logs.val_acc > bestValAcc) { bestValAcc = logs.val_acc; patience = 0; }
        else if (++patience >= 2) { m.stopTraining = true; }
      }
    }
  });

  xs.dispose();
  ys.dispose();

  if (!fs.existsSync(OUT_DIR)) fs.mkdirSync(OUT_DIR);
  await m.save(`file://${OUT_DIR}`);
  console.log(`\nDone! Model saved to ${OUT_DIR}/`);
  console.log('Serve model/ alongside index.html and reload the game.');
}

main().catch(e => { console.error(e); process.exit(1); });

File: train.py
#!/usr/bin/env python3
"""
One-time offline training script.
Run:   python train.py
Output: model/model.json + model/*.bin  (load with tf.loadLayersModel('./model/model.json'))

Requires:  pip install tensorflow tensorflowjs
Data:      data/emnist-letters-train-images-idx3-ubyte
           data/emnist-letters-train-labels-idx1-ubyte
"""

import os
import json
import struct
import numpy as np
import tensorflow as tf

IMAGES_PATH = os.path.join('data', 'emnist-letters-train-images-idx3-ubyte')
LABELS_PATH = os.path.join('data', 'emnist-letters-train-labels-idx1-ubyte')
OUT_DIR     = 'model'
TRAINING_N  = 30000
EPOCHS      = 20


def load_idx3(path):
    with open(path, 'rb') as f:
        magic, count, rows, cols = struct.unpack('>IIII', f.read(16))
        assert magic == 0x00000803, 'Bad IDX3 magic'
        data = np.frombuffer(f.read(), dtype=np.uint8)
    return data.reshape(count, rows, cols), rows, cols


def load_idx1(path):
    with open(path, 'rb') as f:
        magic, count = struct.unpack('>II', f.read(8))
        assert magic == 0x00000801, 'Bad IDX1 magic'
        data = np.frombuffer(f.read(), dtype=np.uint8)
    return data


def fix_orientation(images):
    # EMNIST images are stored transposed — match the JS fix
    return images.transpose(0, 2, 1)


def build_model():
    m = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, 3, activation='relu', padding='same', input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D(2),
        tf.keras.layers.Conv2D(64, 3, activation='relu', padding='same'),
        tf.keras.layers.MaxPooling2D(2),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation='relu'),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Dense(26, activation='softmax'),
    ])
    m.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    return m


def save_tfjs_model(model, output_dir):
    """Serialize a Keras model to TF.js layers-model format (no tensorflowjs needed)."""
    weight_specs = []
    buffers = []
    for w in model.weights:
        arr = w.numpy().astype(np.float32)
        name = w.name.removesuffix(':0')
        weight_specs.append({'name': name, 'shape': list(arr.shape), 'dtype': 'float32'})
        buffers.append(arr.flatten().tobytes())

    weights_file = 'group1-shard1of1.bin'
    with open(os.path.join(output_dir, weights_file), 'wb') as f:
        for buf in buffers:
            f.write(buf)

    manifest = {
        'format': 'layers-model',
        'generatedBy': 'keras',
        'convertedBy': None,
        'modelTopology': json.loads(model.to_json()),
        'weightsManifest': [{'paths': [weights_file], 'weights': weight_specs}],
    }
    with open(os.path.join(output_dir, 'model.json'), 'w') as f:
        json.dump(manifest, f)


def main():
    print('Loading EMNIST data…')
    images, rows, cols = load_idx3(IMAGES_PATH)
    labels = load_idx1(LABELS_PATH)

    n = min(len(images), TRAINING_N)
    images = fix_orientation(images[:n]).astype(np.float32) / 255.0
    images = images[..., np.newaxis]   # add channel dim → (n, 28, 28, 1)
    labels = labels[:n].astype(np.int32) - 1  # shift 1–26 → 0–25

    print(f'Training on {n} samples…')
    model = build_model()
    model.fit(
        images, labels,
        epochs=EPOCHS,
        batch_size=512,
        validation_split=0.1,
        callbacks=[tf.keras.callbacks.EarlyStopping(monitor='val_accuracy', patience=2, restore_best_weights=True)],
    )

    os.makedirs(OUT_DIR, exist_ok=True)
    save_tfjs_model(model, OUT_DIR)
    print(f'\nDone! Model saved to {OUT_DIR}/')
    print('Commit the model/ directory and deploy alongside index.html.')


if __name__ == '__main__':
    main()
